Adds central UI

swiftdiaries opened the pull request:
Adds a Central UI dashboard which provides links to all other components (jobs dashboard, jupyterhub) as a separate component.  
Addresses Issue #146. 

Depends on #11. A secure proxy is needed for the links to point to their respective components.


![screen shot](https://user-images.githubusercontent.com/23024031/35367792-3fe76da0-014e-11e8-8460-bdac5d1f9264.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kubeflow/146)
<!-- Reviewable:end -->

wbuchwalter commented:
Nice 👍 
What would be the steps to deploy this in the cluster?  (I guess you will need to add a new ksonnet template or modify the core one?)

Also we should wait for https://github.com/google/kubeflow/issues/11 before merging this one so that the link can acutally point somewhere.

swiftdiaries commented:
> What would be the steps to deploy this in the cluster? (I guess you will need to add a new ksonnet template or modify the core one?)

I'll modify the core ksonnet. 

> Also we should wait for #11 before merging this one so that the link can acutally point somewhere.

Ah, yes.

Jimexist commented:
@swiftdiaries do you mind putting up a screenshot? Thanks

swiftdiaries commented:
Updated with a screenshot.

jlewi commented:
Can we remove Argo until we add it?

jlewi commented:
/ok-to-test

jlewi commented:
Please update the PR description to describe what this PR is providing.  And if its blocked on #11 please make a note of that as well (I will try to send out the PR with Ambassador today).

jlewi commented:
#186 adds ambassador as a reverse proxy is out for review. We can define services and mappings that the central UI depends on.

#186 serves the TFJobs ui at "/tfjobs/ui/"
https://github.com/kubeflow/kubeflow/pull/166/files
So we should be able to update the UI to point to that.

For tensorboard there will be no single link for tensorboard. The links for tensorboard will be something like
/tensorboard/<name>
where <name> is specific to an instance of tensorboard for a single job.

We should probably remove tensorboard for now and maybe open up an issue for what we want to do about it. One option would be to create a mini UI that would use the K8s API to identify all the K8s services corresponding to TB and then automatically create links to them.

JupyterHub will be served at "/hub/"

K8s dashboard will be at "/k8s/ui/"

swiftdiaries commented:
I've changed links accordingly and removed Tensorboard.
As of now, the central UI can be deployed using docker. 
I am hoping to make a separate PR or add another commit in a few days to add the central UI to the core ksonnet registry.

ScorpioCPH commented on components/centraldashboard/docker/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM alpine
+WORKDIR /app
+COPY --from=build-env /go/bin/centraldashboard /app/
+ENTRYPOINT ./centraldashboard
```
nit: No newline at EOF.

swiftdiaries replied:
No. That is not the official logo. However, I used the same logo that was there in the k8s dashboard for continuity's sake. 

![logo](https://raw.githubusercontent.com/tensorflow/k8s/ed7cae7552bacfa4d693fb8b5975d2e1ec62194a/dashboard/frontend/public/favicon.ico)

ScorpioCPH commented on components/centraldashboard/frontend/index/index.html:
```diff
@@ -0,0 +1,60 @@
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.3/js/bootstrap.min.js" integrity="sha384-a5N7Y/aK3qNeh15eJKGWxsqtnX/wWdSZSKp+81YjTmS15nvnvxKHuzaWwXHDli+4" crossorigin="anonymous"></script>
+  </body>
+</html>
```
ditto

ScorpioCPH commented on components/centraldashboard/frontend/index/styles.css:
```diff
@@ -0,0 +1,224 @@
+  .mdl-demo .mdl-menu__container {
+    z-index: 99;
+  }
+  
```
ditto

swiftdiaries replied:
Edited out the newline. PTAL.

ScorpioCPH replied:
Sorry, do you add a whitespace at line 224?
<img width="356" alt="screen shot 2018-02-01 at 10 23 01" src="https://user-images.githubusercontent.com/5319646/35657941-02839276-073a-11e8-9b1f-625524c1c020.png">

wbuchwalter commented:
@ScorpioCPH @swiftdiaries I don't think you should use the same logo as the one I did forTfJob dashboard. This logo was just a merge of K8s and TF logos, which is okay for TfJob dashboard as a placeholder while we get something better, but Kubeflow shouldn't be TF specific.
I would rather not have any favicon until we get an official one than this one, as it may introduce confusion regarding what is Kubeflow.

Jimexist commented:
maybe use a brainy thing upon the kubernetes logo?

swiftdiaries commented:
Yeah, that makes sense. I'll remove the favicon for now. Once, we have a Kubeflow org logo, we can update it to include that.

swiftdiaries commented:
@Jimexist That is a great idea.

jlewi commented:
Can you open up an issue to start tracking all the work to make this ready for a 0.1 release

e.g.
1. Integration into ksonnet components
2. end to end tests
3. release process

@Jimexist or @wbuchwalter as the resident UI experts can one of you approve?

swiftdiaries commented:
Ok, I'll start an issue for the 0.1 release.

Jimexist commented on components/centraldashboard/frontend/index/index.html:
```diff
@@ -0,0 +1,58 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">
+    <title>Kubeflow</title>
+
+    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:regular,bold,italic,thin,light,bolditalic,black,medium&amp;lang=en">
```
People in China have problem accessing `googleapis.com` without VPN, so I think a better way is to use `parcel` or other thing to pre-pack the data, but that's a lot more work so I think it's sufficient to stay this way for now.

swiftdiaries commented:
Ill add it to the 0.1 release TODOs and follow it up.

On Fri, Feb 2, 2018 at 1:35 AM Jiayu Liu <notifications@github.com> wrote:

> *@Jimexist* commented on this pull request.
> ------------------------------
>
> In components/centraldashboard/frontend/index/index.html
> <https://github.com/kubeflow/kubeflow/pull/146#discussion_r165567714>:
>
> > @@ -0,0 +1,58 @@
> +<!doctype html>
> +<html lang="en">
> +  <head>
> +    <!-- Required meta tags -->
> +    <meta charset="utf-8">
> +    <meta http-equiv="X-UA-Compatible" content="IE=edge">
> +    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">
> +    <title>Kubeflow</title>
> +
> +    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:regular,bold,italic,thin,light,bolditalic,black,medium&amp;lang=en">
>
> People in China have problem accessing googleapis.com without VPN, so I
> think a better way is to use parcel or other thing to pre-pack the data,
> but that's a lot more work so I think it's sufficient to stay this way for
> now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/kubeflow/pull/146#pullrequestreview-93529781>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV9Rn7xVGVyBZ6cc9JYXpK9ovjhkCOshks5tQqyngaJpZM4Rrrcs>
> .
>

jlewi commented:
@Jimexist @wbuchwalter @swiftdiaries What's the status of this PR?

swiftdiaries commented:
@wbuchwalter thought it was best to add the dash as a ksonnet component to this PR and then test it. So it would be just one PR with everything, instead of having 2 separate PRs as originally intended.

jlewi commented:
@swiftdiaries Any idea when the changes mentioned above will be ready?

swiftdiaries commented:
I have a yaml deployment ready. I'm working on the ksonnet component. I hope to get it done by end of day today.

Jimexist commented:
@swiftdiaries also by the way could you rebase please?

jlewi commented:
@swiftdiaries Do you have an ETA for this PR? Any chance it will be ready by EOW?

swiftdiaries commented:
Yes, it should be ready by the EOW. I've added the dash as a separate ksonnet component. I'm updating the centraldash README and the user_guide on launching the dashboard. I have little progress on https://github.com/kubeflow/tf-operator/issues/323, so I'm trying to work on that now.

swiftdiaries commented:
I've added setup instructions at https://github.com/swiftdiaries/kubeflow/tree/origin/kubeflow/centraldash.
For testing,
```commandline
#Assuming you have a ksonnet-app
# cd into the app directory
$ ks registry add https://github.com/swiftdiaries/kubeflow/tree/origin/kubeflow
$ ks pkg install kubeflow/centraldash
$ NAMESPACE=kubeflow
$ kubectl create namespace ${NAMESPACE}
$ ks generate core kubeflow-core --name=kubeflow-core --namespace=${NAMESPACE}
$ ks generate centraldash kubeflow-centraldash --name=kubeflow-dash --namespace=${NAMESPACE}
# Apply to server.
$ ks apply ${KF_ENV} -c kubeflow-core
$ ks apply ${KF_ENV} -c kubeflow-centraldash
# Centraldash uses the Ambassador service to connect to other UI services.
# Port 4004 is hard-coded for Ambassador's endpoint on centraldash
$ PORT=4004
$ PODNAME=`kubectl get pods --namespace ${NAMESPACE} --selector="service=ambassador" -o \
jsonpath='{.items[0].metadata.name}'`
$ kubectl port-forward --namespace=${NAMESPACE} $PODNAME ${PORT}:80
# In another terminal, start the centraldash service
$ DASHPODNAME=`kubectl get pods -n ${NAMESPACE} --selector="app=centraldash" -o \
jsonpath='{.items[0].metadata.name}'`
$ kubectl port-forward —namespace ${NAMESPACE} ${DASHPODNAME} 8082
```

swiftdiaries commented:
/cc @jlewi @wbuchwalter

wbuchwalter commented:
I'll take a look later today. Thanks.

wbuchwalter commented:
As discussed on slack, could we serve the central UI on ambassador directly by adding a new prefix?
That way we should be able to have this working without needing to port-forward twice?

Other things:
* Can we rename `centraldash` to `centraldashboard` everywhere to make things clearer?
* Reduce the width of the header so it's consistent with k8s dashboard and the dashboard

jlewi commented:
If this is ready for review please remove [WIP] from the title.

@wbuchwalter  can you explain your comment? Why is port-forwarding involved here?

wbuchwalter commented:
@jlewi I am talking about this part:

```sh
# Centraldash uses the Ambassador service to connect to other UI services.
# Port 4004 is hard-coded for Ambassador's endpoint on centraldash
$ PORT=4004
$ PODNAME=`kubectl get pods --namespace ${NAMESPACE} --selector="service=ambassador" -o \
jsonpath='{.items[0].metadata.name}'`
$ kubectl port-forward --namespace=${NAMESPACE} $PODNAME ${PORT}:80
# In another terminal, start the centraldash service
$ DASHPODNAME=`kubectl get pods -n ${NAMESPACE} --selector="app=centraldash" -o \
jsonpath='{.items[0].metadata.name}'`
$ kubectl port-forward —namespace ${NAMESPACE} ${DASHPODNAME} 8082
```
I am just discovering ambassador on my side, so what I say might be wrong.
But assuming I have all the component and ambassador running in my cluster, I should be able to port-forward ambassador (or use a LoadBalancer, or k8s proxy) and from there access the central ui on some prefix such as `/centralui`.
Why should we make users port-forward the central ui pod?

jlewi commented:
Review status: 0 of 10 files reviewed at latest revision, 11 unresolved discussions.

---

*[components/centraldashboard/config/Dockerfile, line 3 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8FwGjUFbx4XrWITv:-L6s8FwGjUFbx4XrWITw:b890jak) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/components/centraldashboard/config/Dockerfile#L3)):*
> ```Dockerfile
> # build prod
> FROM golang:alpine
> LABEL maintainer="Adhita Selvaraj <adhita.selvaraj@gmail.com>"
> ```

Please file an issue to add this to our release process.

---

*[components/centraldashboard/config/frontend.yaml, line 2 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8LuLSQiZgo3y06eo:-L6s8LuLSQiZgo3y06ep:b-qat0j3) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/components/centraldashboard/config/frontend.yaml#L2)):*
> ```YAML
> ---
> apiVersion: extensions/v1beta1
> ```

What is this YAML file for? Isn't centraldash deployed via ksonnet?

---

*[components/centraldashboard/frontend/index/index.html, line 10 at r6](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L5GFSx1hGsHJETz79db-r6-10:-L6s8UoieBq2oBPfkW1t:b-kh4b1m) ([raw file](https://github.com/kubeflow/kubeflow/blob/cea008548ace32342be5a99eaaddcde945729cb6/components/centraldashboard/frontend/index/index.html#L10)):*
<details><summary><i>Previously, Jimexist (Jiayu Liu) wrote…</i></summary><blockquote>

People in China have problem accessing `googleapis.com` without VPN, so I think a better way is to use `parcel` or other thing to pre-pack the data, but that's a lot more work so I think it's sufficient to stay this way for now.
</blockquote></details>

I'd suggest filing a follow up issue.

---

*[kubeflow/registry.yaml, line 16 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8acOdbxchPum6ga7:-L6s8acOdbxchPum6ga8:b-kxucpr) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/registry.yaml#L16)):*
> ```YAML
>     version: master
>     path: argo
>   centraldash:
> ```

Can we just put the ksonnet in the existing core package rather than creating a new package?

---

*[kubeflow/centraldash/centraldash.libsonnet, line 113 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8krjv4nWG9B7P-Yx:-L6s8krjv4nWG9B7P-Yy:bn9mpzn) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/centraldash.libsonnet#L113)):*
>  Is this comment a left over?

---

*[kubeflow/centraldash/README.md, line 3 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s90kT5hRRf7Ah0StS:-L6s90kT5hRRf7Ah0StT:b-6xvyx5) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/README.md#L3)):*
> ```Markdown
> # Centraldash
> 
> > Ways for deploying Centraldash
> ```

Central dash should probably be part of the kubeflow-core component. Since this is still experimental; I'd suggest adding an option to the existing prototype
https://github.com/kubeflow/kubeflow/blob/master/kubeflow/core/prototypes/all.jsonnet

That enables the centraldash or not and default to false. 

---

*[kubeflow/centraldash/README.md, line 30 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s9ht7-LGyFbPIUn77:-L6s9ht7-LGyFbPIUn78:b-mjdlr6) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/README.md#L30)):*
> ```Markdown
> $ ks apply ${KF_ENV} -c kubeflow-centraldash
> # Centraldash uses the Ambassador service to connect to other UI services.
> # Port 4004 is hard-coded for Ambassador's endpoint on centraldash
> ```

I don't understand this? Can you point me to the relevant location?
Ambassador should be configured as follows some prefix like
"/central-dash"
redirects traffic to the K8s service that since in front of the central UI.

Ambassador should also have its lowest priority route for "/" direct to the central UI
so that by default the URL
http://<doman>/ gets redirected to central UI

---

*[kubeflow/centraldash/README.md, line 38 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6sANrloTGp6SOUaFDR:-L6sANrma277Uzlf16r3:bbuol55) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/README.md#L38)):*
> ```Markdown
> $ DASHPODNAME=`kubectl get pods -n ${NAMESPACE} --selector="app=centraldash" -o \
> jsonpath='{.items[0].metadata.name}'`
> $ kubectl port-forward —namespace ${NAMESPACE} ${DASHPODNAME} 8082
> ```

Why is this needed? My expectation is that one should be able to connect to the Ambassador pod and from there you should be able to access all URLs at the relevant prefixes

For example
kubectl port-forward ${ambassador pod} 8888:<ambassador port>

http://localhost:8888/tensoboard/<job name>/
http://localhost:8888/tf-jobs-ui/
http://localhost:8888/central-dash/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[components/centraldashboard/config/frontend.yaml, line 2 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8LuLSQiZgo3y06eo:-L6sHW_2CeKpOchUZhB5:b-tf8sam) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/components/centraldashboard/config/frontend.yaml#L2)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

What is this YAML file for? Isn't centraldash deployed via ksonnet?
</blockquote></details>

I was initially experimenting with a YAML file. I'll remove it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[components/centraldashboard/frontend/index/index.html, line 10 at r6](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L5GFSx1hGsHJETz79db-r6-10:-L6sHv6__V-7DmxbYF-k:bloxtva) ([raw file](https://github.com/kubeflow/kubeflow/blob/cea008548ace32342be5a99eaaddcde945729cb6/components/centraldashboard/frontend/index/index.html#L10)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

I'd suggest filing a follow up issue.
</blockquote></details>

I've filed a tracking progress in this: https://github.com/kubeflow/kubeflow/issues/199

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[kubeflow/registry.yaml, line 16 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8acOdbxchPum6ga7:-L6sI89Fa59OV8X5okPj:bdwk44w) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/registry.yaml#L16)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Can we just put the ksonnet in the existing core package rather than creating a new package?
</blockquote></details>

Sure thing. I'll restructure it into the core component.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[kubeflow/centraldash/centraldash.libsonnet, line 113 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s8krjv4nWG9B7P-Yx:-L6sIHBQiR5FDu2rkJ4x:b-avcr2k) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/centraldash.libsonnet#L113)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

>  Is this comment a left over?
</blockquote></details>

Yes. I used the Argo component as the base. I'll remove it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[kubeflow/centraldash/README.md, line 3 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s90kT5hRRf7Ah0StS:-L6sIR-Yne8rQ2sf8kfO:b-4hvi0m) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/README.md#L3)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Central dash should probably be part of the kubeflow-core component. Since this is still experimental; I'd suggest adding an option to the existing prototype
https://github.com/kubeflow/kubeflow/blob/master/kubeflow/core/prototypes/all.jsonnet

That enables the centraldash or not and default to false. 
</blockquote></details>

Got it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

swiftdiaries commented:
*[kubeflow/centraldash/README.md, line 30 at r8](https://reviewable.io:443/reviews/kubeflow/kubeflow/146#-L6s9ht7-LGyFbPIUn77:-L6sIUF1g_1bjKanSnkI:b-ylbsor) ([raw file](https://github.com/kubeflow/kubeflow/blob/8d2c542198265740d4dfca20d68c6ade3eaa8d21/kubeflow/centraldash/README.md#L30)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

I don't understand this? Can you point me to the relevant location?
Ambassador should be configured as follows some prefix like
"/central-dash"
redirects traffic to the K8s service that since in front of the central UI.

Ambassador should also have its lowest priority route for "/" direct to the central UI
so that by default the URL
http://<doman>/ gets redirected to central UI
</blockquote></details>

I was under the impression that this would be a separate component. I'll add it as a ambassador route and simplify the setup process.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/kubeflow/146)*
<!-- Sent from Reviewable.io -->

jlewi commented:
@swiftdiaries Thanks let me know when this is ready for another look.

@wbuchwalter @swiftdiaries Would it make sense to rethink splitting this PR and moving the ksonnet configs into a new PR? I think the central UI is ready or almost ready? So it could probably be submitted now.

wbuchwalter commented:
@jlewi I am ok with splitting this if it makes it easier.
I would just like to the two nits mentioned before addressed before merging however:
> * Can we rename centraldash to centraldashboard everywhere to make things clearer?
> * Reduce the width of the header so it's consistent with k8s dashboard and the tfjob dashboard

swiftdiaries commented:
Sorry for the late reply. I'll address these 2 nits and then merge before addressing the remaining issues in a later PR.

jlewi commented:
@swiftdiaries Any chance we can get the above issues addressed and this PR committed this week?

swiftdiaries commented:
Sorry for the late reply. I'll try to get in the changes by the end of the week.

jlewi commented:
@swiftdiaries Could we get this PR updated and committed?

It looks like the changes needed are relatively minor

1. Delete the YAMl file
1. Move the ksonnet components into the core package.

swiftdiaries commented:
Closing in place of a new PR.
